### ISSUE TITLE
V4 - switch to Stylus and BEM

### ISSUE THREAD

**Issue by piupiupiu** [twbs/bootstrap#17484]
Hello guys.

I have read your announce about TWBT4 and excited to see the beta version.
However, as many others, I'm concerned about the technologies/approaches you use in it. As when it is released, it will be outdated from the first day and looks like rebuilding Fiat 124 from the scratch in 2015. Please let me explain:
– Why did you decide to move from LESS to SASS and NOT straight to Stylus? It looks like trade bad for worse. Both SASS and LESS and tremendous outdated monsters. Why did not you decide to use Stylus instead? I, as many other front-end developers, was an apologist of SASS for a long period of time unless I finally tired of wasting time and glitches of it. So, I moved to Stylus and now the things are working much-much faster. Yes, still lots of people use SASS. Just because they don't know about Stylus.
– Why don't you use BEM (https://en.bem.info/)? You have a great competitor - material design framework by google which already use this approach. I'm not here to describe all the benefits of it but you should consider using it.
You, guys, are doing a great work, however, I think, it is necessary to keep up with the times and follow all modern approaches and technologies popular within frontend community. What do you think?
Probably, you have strong reasons for that. If so, could you, please, describe your points?
Thanks in advance.

**Comment by cvrebert** [comment]
I believe mdo dislikes Stylus for the same reason that he opted to use SCSS instead of Sass' indented syntax: The basic syntax is too different from that of CSS.

**Comment by cvrebert** [comment]
As for BEM, I don't think @mdo's opinion has changed since #9366.

**Comment by mdo** [comment]
Yeah nothing's changed here. Stylus isn't anything revolutionary and BEM doesn't have much of an impact on anything beyond class names.

**Comment by piupiupiu** [comment]
>BEM doesn't have much of an impact on anything beyond class names.

I completely disagree with you. BEM is not just about naming. And I think you did not check it out thoroughly. And please, re-open the issue. Please, let's others discuss this problem.

First of all it's about the performance. 

Let's take a look at `_pagination.scss`. What we see there:

    .pagination {
        > li {
            > a,
            > span {

The result CSS will look as follows:

    .pagination > li > a,
    .pagination > li > span

How browser will understand this structure? CSS selectors are processed from RIGHT to LEFT.
So, It will take all (ALL, Carl!) anchors. Then check which of them are direct children of `li` tags and then check which of them are direct children of `.pagination` class.

This is MUCH much slower than operating with class names.

You may say it does not matter, modern browsers and hardware is fast enough blah-blah-blah. 
I will disagree with you here as well.
It does not matter in case of a pet project. In case of real big site with hundreds of links/spans/divs etc it matters.

Refer to prof link on Mozilla Developer Network:

https://developer.mozilla.org/en-US/docs/Web/Guide/CSS/Writing_efficient_CSS#How_the_Style_System_Matches_Rules

I'm not going to copy-paste here the other benefits, just please refer to the article:

http://mono.company/journal/frontend/learning-to-love-bem/

**Comment by piupiupiu** [comment]
>I believe mdo dislikes Stylus for the same reason that he opted to use SCSS instead of Sass' indented >syntax: The basic syntax is too different from that of CSS.

Yes, it's too different. It's much shorter and neat. Why are you still trying to stay sticky to CSS-syntax?
Well, use the pure CSS in this case, instead of pre-processors. But think over the millions of hours saved for millions of developers world-wide. It's not about syntax after all. It's about performance.
SASS is SLOW. Even on my iMac 3.2GHz with 32GB of RAM it takes up to 5-10 seconds to compile a really big project. After wasted hours. I migrated to project to Stylus and now the project is compiled within 100-300ms. I may love or not love Stylus. But think about it. Millions of developers will say 'thank you' to you.

**Comment by cvrebert** [comment]
> This is MUCH much slower than operating with class names.

You're correct that the pagination component ought to use classes instead of element selectors and child selectors. The fact that it currently doesn't is a bug that we ought to fix. In fact, using classes instead of ultra-specific selectors like those is one of our action items for v4. But v4 is a work-in-progress and we haven't gotten around to fixing this everywhere yet.

So basically, this particular example isn't indicative of our CSS style for v4, which is similar to BEM.

> Why are you still trying to stay sticky to CSS-syntax?

Because folks are already familiar with it and it's more obvious what the final output CSS will look like. And, when necessary, you can mostly just copy-paste in CSS from elsewhere without having to translate it into a very different syntax.

> SASS is SLOW

I am guessing you haven't tried [libsass](https://github.com/sass/libsass) yet? libsass is fast. *Ruby Sass* is indeed slow.
This is why we give the user the option to choose which one of them to use when compiling Bootstrap.

**Comment by piupiupiu** [comment]
>I am guessing you haven't tried libsass yet? libsass is fast. Ruby Sass is indeed slow.
I tried it. I tried all existing ways to compile SASS. Libsass is the same slow thing. May be 10-20% faster that rubysass, but still dramatically slower than Stylus.
And speaking about libsass. Haven't you met any bugs with it which compiled into improper CSS? 
You will meet then in the future for sure. I cannot provide you with the cases now, as I moved to Stylus and forgot about any troubles. But that was a project based on twbt V3 and Sass (scss-syntax).

**Comment by mdo** [comment]
I misspoke when I said class names earlier—I should've said selectors in general. I stand by that though—there's far more to consider than your selectors when building a toolkit like Bootstrap that's used by millions of folks.

At any rate, there are a number of things to consider here that you're overlooking.

- We emphasize source code that is as close to the compiled CSS as possible. We could omit curly braces and semi-colons, but we'd gain nothing. They'd still be back when you compiled it.
- You're suggesting Stylus is superior to these other preprocessors, but you don't present any factual data to back that up. Instead you insult our decisions with absolutely no context.
- You inferred my comment on class names meant anything about hardware or browser rendering speeds. I didn't say anything of the sort, and I don't disagree with you on one part—classes are much faster, but they're not the only concern with CSS.
- We've already committed to Sass for v4. I doubt folks out there will be saying thank you to us for switching mid-development to yet another preprocessor that a small percent of folks use by comparison.
- What about linters for Stylus? 

Addressing your latest comment:

> I tried all existing ways to compile SASS. Libsass is the same slow thing. May be 10-20% faster that rubysass, but still dramatically slower than Stylus.

Libsass is far faster than Ruby Sass (and even Less in my experience moving to Sass in v4). See an example benchmarking post: http://www.solitr.com/blog/2014/01/css-preprocessor-benchmark/. Do you have any examples you can point to regarding the speed of Stylus?

> And speaking about libsass. Haven't you met any bugs with it which compiled into improper CSS?

I've encountered incorrect compiled CSS with LibSass, Ruby Sass, and Less. Using a preprocessor does not automatically stop one from writing invalid CSS.

> You will meet then in the future for sure. I cannot provide you with the cases now, as I moved to Stylus and forgot about any troubles.

Paranoia and vague allusions to problems we haven't encountered isn't a helpful comment. What problems will we run into, and that we haven't yet already resolved? Can you provide cases or examples for any of this?

---

Bottom line, you've presented no data on benefits of Stylus. If we wanted a "neater" syntax, we could use the `.sass` syntax (instead of `.scss`) which allows for that. We already see a significantly faster compilation time for the Bootstrap core: the `grunt sass:core` task on our v4 branch compiles our CSS in `328ms`. Does Stylus do that much better while providing anything of value that Sass doesn't?

**Comment by felixfbecker** [comment]
Sass does not resolve relative urls when you @import .scss files. Which means, when I import bootstrap into my styles, the @font-face url() declarations will be broken. I know glyphicons are being removed but the same applies to font-awesome. I have to choose one preprocessor and if I use Bootstrap then I should use Sass. Stylus, Less, they all have an option like --relative-url that fixes urls when importing. This is very important for a framework like bs thats installed with package managers. I know I can set some variables but I would have to do that in my .scss. I would rather just do gulp build and let Sass take care of it. The Sass devs rejected this multiple times because they expect it to be fixed by higher-level frameworks like Compass with asset helpers.

**Comment by felixfbecker** [comment]
I mean, I dont like the language either (even though they say they support vanilla CSS), just as I dont like CoffeeScript, but this is a dealbreaker for Sass.
